### PR TITLE
Allow transfer codings before final chunked on HTTP/1 send

### PR DESCRIPTION
**Pull request by Kludex**
## The bug

The uvicorn zttp branch found a framing mismatch around non-chunked transfer
codings. The same assumption existed inside zttp 0.0.16:

```python
c.send_response(200, [(b"Transfer-Encoding", b"gzip")])
c.send_data(b"hello")
c.end_message()
```

produced:

```http
HTTP/1.1 200 OK
Transfer-Encoding: gzip

5
hello
0


```

The header declares only `gzip`, while the body is chunk-framed. A client follows
the declared coding order and tries to gunzip the chunk framing itself.

#156 made this safe in 0.0.17 by rejecting every send-side
`Transfer-Encoding` except one literal `chunked` field. That prevents corrupt
wire output, but is narrower than RFC 9112 and narrower than zttp's own reader:

| Header(s) | Reader | Writer before this PR |
| --- | --- | --- |
| `chunked` | accept | accept |
| `gzip, chunked` | accept | reject |
| `gzip` + `chunked` (two field-lines) | accept | reject |
| `gzip` | reject | reject |
| `chunked, gzip` | reject | reject |

This matters for uvicorn's fix: it appends `chunked` to an existing transfer-
coding list. Both RFC-valid representations it can produce were rejected by
zttp 0.0.17.

## The fix

Move Transfer-Encoding validation into one helper in `framing.zig`, shared by
both read and write. Both paths now follow the same rule:

1. Fold every `Transfer-Encoding` field-line into one ordered comma list (RFC
   9110 §5.2 / RFC 9112 §6.1).
2. Require `chunked` exactly once and last.
3. Let codings before it describe transformations the caller has already
   applied; zttp supplies only the final chunk framing it implements.

So these now work:

```http
Transfer-Encoding: gzip, chunked
```

and the equivalent split form:

```http
Transfer-Encoding: gzip
Transfer-Encoding: chunked
```

The supplied field-lines are preserved on the wire; only their combined ordering
is validated.

## Security invariants retained

This is not a general Transfer-Encoding relaxation. All ambiguous/unframeable
forms remain `LocalProtocolError`:

- `gzip` — no final framing coding
- `chunked, gzip` — coding after `chunked`
- `chunked, chunked` — duplicate `chunked`
- split `chunked` + `chunked` — duplicate across field-lines
- any `Transfer-Encoding` together with `Content-Length`

This also answers the "append if it does not end in chunked" suggestion from the
uvicorn review: zttp never repairs `chunked, gzip` into
`chunked, gzip, chunked`. It rejects the malformed input, and never declares
`chunked` twice.

## Caller responsibility

zttp does not apply `gzip` (or any preceding coding). `send_data()` must already
contain gzip-compressed bytes; zttp only wraps those bytes in chunks. The HTTP/1
documentation now says this explicitly and gives a `gzip.compress()` example.

## Tests

- Exact-wire Zig tests for a single `gzip, chunked` field and split `gzip` +
  `chunked` fields.
- Public Python tests for both accepted forms.
- Public Python regressions for no-final-chunked, non-final chunked, and duplicate
  chunked both within and across field-lines.
- The executable documentation caught and prevented state leaking into the next
  example.

Validation:

- `zig build test`: **807/807**
- `scripts/check`: clean
- `scripts/test`: **313 passed**, 1 optional interop skip
- coverage: **100.00%**
- `scripts/docs`: `No issues found` under strict mode

---
## Summary by cubic
HTTP/1 reader and writer now accept transfer codings before a final `chunked`, strictly validate `Transfer-Encoding` grammar, and reject TE where HTTP forbids it (1xx/204 and successful CONNECT), aligning both directions with RFC 9112/9110. This fixes framing mismatches, prevents misparsing on inbound bodyless responses, and unblocks `uvicorn`; docs and tests use valid gzip bytes.

- **Bug Fixes**
  - Accept `Transfer-Encoding: gzip, chunked` and the split form; preserve field-lines and emit only chunk framing.
  - Validate full grammar: token names/parameters and quoted strings; reject empty list members on send, ignore on receive (RFC 9110 §5.6.1).
  - Forbid TE on `1xx`/`204` and successful `CONNECT` on send and receive; allow `HEAD`/`304` to carry metadata; fail before writing any bytes.
  - Keep rejecting: `gzip` alone, `chunked, gzip`, duplicate `chunked` (including across lines), any TE with `Content-Length`, and parameters on `chunked`.

- **Refactors**
  - Centralized TE checks in `framing.validateTransferEncoding` with distinct `receive`/`send` modes and a new `forbid_transfer_encoding` framing option used by reader and writer.
  - Docs clarify callers must pre-encode bytes (`gzip.compress(..., mtime=0)`); tests add exact-wire cases, malformed grammar, and forbidden-context regressions.

<sup>Written for commit 1a97276e02a5ebff9ba7913f5872001fb3bde39d. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/Kludex/zttp/pull/170?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Supports ordered `Transfer-Encoding` where `chunked` appears exactly once and last, including `gzip, chunked`, and supports `Transfer-Encoding` split across multiple header fields.
* **Bug Fixes**
  * Rejects malformed/invalid `Transfer-Encoding` (duplicate/misordered `chunked`, invalid syntax/parameters) and still rejects ambiguous `Transfer-Encoding` + `Content-Length`.
  * Enforces response rules that forbid `Transfer-Encoding` for `1xx`, `204`, successful `CONNECT`, and informational responses.
  * Ensures bodyless responses can retain `Transfer-Encoding` metadata.
* **Documentation**
  * Updates HTTP/1.1 guidance with clarified framing rules and examples.
* **Tests**
  * Adds/updates reader/writer coverage for acceptance, rejection, and edge cases (including parameter/quoted-comma handling).

A follow-up reader review found the matching inbound gap (`1a97276`): 204 and
successful CONNECT responses carrying TE were accepted as bodyless, so following
chunk bytes could be reinterpreted as a new message. The reader now derives an
explicit `forbid_transfer_encoding` option from the remembered method and status
before framing. HEAD/304 retain their permitted metadata case; informational
responses were already rejected by their dedicated path.

### CODE REVIEW

File: tests/test_write.py [hunk 1]
```diff
@@ -97,6 +97,39 @@ def test_chunked_response() -> None:
...
+)
+def test_chunked_framing_with_preceding_transfer_codings(headers: list[tuple[bytes, bytes]]) -> None:
+    conn = zttp.Connection(zttp.SERVER)
+    conn.send_response(200, headers)
+    conn.send_data(b"compressed")
+    conn.end_message()
+
+    head, body = conn.data_to_send().split(b"\r\n\r\n", 1)
+    assert head.count(b"chunked") == 1
+    assert body == b"a\r\ncompressed\r\n0\r\n\r\n"
```
[the remark this answers is not in the document]
Comment (Kludex) [reply tests/test_write.py]: Good catch — fixed in 5c9780f.

- Both exact-wire Zig tests now send one fixed, deterministic gzip member for `hello` (mtime zero), and the expected chunk size/wire bytes match it.
- The Python test uses `gzip.compress(b"hello", mtime=0)`, builds the expected chunk from the actual compressed length, and decompresses the fixture back to `hello` so the declared coding and bytes cannot drift apart.

Re-verified: 807/807 Zig tests, 313 Python tests, 100% coverage, full checks clean.

File: src/core/h1/writer.zig [hunk 2]
```diff
@@ -179,21 +179,16 @@ fn validateHeaders(hdrs: []const Header) WriteError!void {
...
-/// conflicting duplicate Content-Lengths would let a downstream parser disagree
-/// about message boundaries (response splitting).
+/// smuggling guards. Transfer-Encoding field-lines form one ordered comma list;
+/// codings may precede `chunked`, but `chunked` must appear exactly once and last.
+/// The caller supplies bytes to which those preceding codings are already applied;
+/// zttp supplies only the final chunk framing. TE + CL and conflicting duplicate
+/// Content-Lengths remain errors.
 fn validateFraming(hdrs: []const Header) WriteError!void {
-    var has_te = false;
+    const has_te = framing.validateTransferEncoding(hdrs) catch return error.InvalidField;
```
[the remark this answers is not in the document]
Comment (Kludex) [reply src/core/h1/writer.zig]: Valid — fixed in fc74f80, on both send and receive rather than with a writer-only precheck.

The accumulator now parses the actual `#transfer-coding` grammar: token coding names, token or quoted-string parameter values, quoted-pair, OWS, and separators. These now fail before any output:

- `gzip,,chunked`, `,chunked`, `chunked,` on send
- `;bad, chunked`
- `gzip;flag, chunked` (missing `=`)
- empty parameter values / unterminated quoted strings
- parameters on `chunked` (preserves the reader's prior exact-chunked policy)

There is one direction-specific RFC nuance: senders MUST NOT generate empty list members, while recipients MUST ignore a reasonable number (RFC 9110 §5.6.1). The shared validator therefore takes an explicit `.send` / `.receive` mode. `gzip,,chunked` remains accepted inbound but is rejected outbound; malformed coding/parameter syntax is rejected in both directions.

Added core and public regressions; 814/814 core tests and the full suite pass.

[the remark this answers is not in the document]
Comment (Kludex) [reply src/core/h1/writer.zig]: Valid, and partly pre-existing: even a plain `Transfer-Encoding: chunked` could be emitted on 204/1xx/2xx CONNECT before this PR. Fixed in fc74f80 at the status/method boundary, before `writeStatusLine`, so an error leaves the pending output empty.

Rejected now:

- any 1xx response through `send_response`
- interim 1xx through `send_informational`
- 204
- successful CONNECT

HEAD and 304 deliberately remain allowed: RFC 9110 permits them to describe the coding that would have applied to the corresponding body. Tests cover both permitted exceptions and all forbidden contexts.

File: src/core/h1/framing.zig [hunk 3]
```diff
@@ -67,6 +174,14 @@ pub fn responseIsBodyless(method: []const u8, status: u16) bool {
     return false;
 }
 
+/// RFC 9112 6.1 forbids Transfer-Encoding on 1xx/204 and successful CONNECT
+/// responses. HEAD and 304 are also bodyless, but may carry the coding that
+/// would have applied to the corresponding body (RFC 9110 8.6, 15.4.5).
+pub fn responseForbidsTransferEncoding(method: []const u8, status: u16) bool {
```
[the remark this answers is not in the document]
Comment (Kludex) [reply src/core/h1/framing.zig:180]: Valid — fixed in 1a97276.

Reproduced before the fix:

- inbound 204 + `Transfer-Encoding: chunked` → Response + EndOfMessage
- inbound 200 to CONNECT + TE → Response + EndOfMessage
- HEAD/304 + `gzip, chunked` → bodyless metadata (legal)
- 103 + TE was already rejected by the separate informational-response path

The framing API now has an explicit `forbid_transfer_encoding` option separate from `bodyless`. The reader derives both from the remembered request method and response status before `frameBody()`: 204 and successful CONNECT reject TE, while HEAD and 304 retain the RFC-permitted metadata case. This prevents chunk bytes after an illegal bodyless head from being reinterpreted as the next message.

Added core and public tests for both forbidden contexts and both permitted exceptions. Final validation: 817/817 Zig tests, 327 Python tests, 100% coverage, strict docs and full checks clean.

**Comment by Kludex** [comment]
Closing in favor of the smaller h11-compatible policy.

zttp 0.0.17 already fixes the original corrupt-wire bug by rejecting every send-side `Transfer-Encoding` except literal `chunked`, exactly like h11. Supporting `gzip, chunked` would add transfer-coding grammar and caller-contract surface for a coding zttp does not perform; uvicorn should reject unsupported app-supplied transfer codings rather than ask zttp to preserve them.

The independent bodyless-response bug found during review is extracted to #171: TE was accepted/emitted on 1xx, 204, and successful CONNECT, and inbound chunk bytes could be reinterpreted as a next message. #171 fixes only that, preserving the existing narrow TE policy.
